exercises/react/test/test_react.c

```diff
@@ -0,0 +1,295 @@
+   free_reactor(r);
+}
+
+static void cb_noop(void *obj __attribute__ ((unused)), int v
+                    __attribute__ ((unused)))
+{
+}
+
+void test_compute_cells_dont_access_callback_obj(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 1);
+   struct cell *output = create_compute1_cell(r, input, plus1);
+
+   add_callback(output, NULL, cb_noop);
+
+   cell_value_set(input, 3);
+
+   free_reactor(r);
+}
+
+static int big_if_three(int x)
+{
+   return x < 3 ? 111 : 222;
+}
+
+void test_callbacks_only_fire_on_change(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 1);
+   struct cell *output = create_compute1_cell(r, input, big_if_three);
+
+   struct cbinfo cbinfo = { -1, 0 };
+   add_callback(output, &cbinfo, cb_record_last_value);
+
+   cell_value_set(input, 2);
+   TEST_ASSERT_EQUAL_INT(0, cbinfo.times_called);
+
+   cell_value_set(input, 4);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo.times_called);
+   TEST_ASSERT_EQUAL_INT(222, cbinfo.last_value);
+
+   free_reactor(r);
+}
+
+void test_callbacks_can_be_added_and_removed(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 11);
+   struct cell *output = create_compute1_cell(r, input, plus1);
+
+   struct cbinfo cbinfo1 = { -1, 0 };
+   callback_id cb1 = add_callback(output, &cbinfo1, cb_record_last_value);
+   struct cbinfo cbinfo2 = { -1, 0 };
+   add_callback(output, &cbinfo2, cb_record_last_value);
+
+   cell_value_set(input, 31);
+
+   remove_callback(output, cb1);
+   struct cbinfo cbinfo3 = { -1, 0 };
+   add_callback(output, &cbinfo3, cb_record_last_value);
+
+   cell_value_set(input, 41);
+
+   TEST_ASSERT_EQUAL_INT(1, cbinfo1.times_called);
+   TEST_ASSERT_EQUAL_INT(32, cbinfo1.last_value);
+   TEST_ASSERT_EQUAL_INT(2, cbinfo2.times_called);
+   TEST_ASSERT_EQUAL_INT(42, cbinfo2.last_value);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo3.times_called);
+   TEST_ASSERT_EQUAL_INT(42, cbinfo3.last_value);
+
+   free_reactor(r);
+}
+
+void test_removing_a_callback_multiple_times_doesnt_interfere_with_others(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 11);
+   struct cell *output = create_compute1_cell(r, input, plus1);
+
+   struct cbinfo cbinfo1 = { -1, 0 };
+   callback_id cb1 = add_callback(output, &cbinfo1, cb_record_last_value);
+   struct cbinfo cbinfo2 = { -1, 0 };
+   add_callback(output, &cbinfo2, cb_record_last_value);
+   for (int i = 0; i < 10; ++i) {
+      remove_callback(output, cb1);
+   }
+
+   cell_value_set(input, 2);
+
+   TEST_ASSERT_EQUAL_INT(0, cbinfo1.times_called);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo2.times_called);
+   TEST_ASSERT_EQUAL_INT(3, cbinfo2.last_value);
+
+   free_reactor(r);
+}
+
+static int minus1(int x)
+{
+   return x - 1;
+}
+
+static int times(int x, int y)
+{
+   return x * y;
+}
+
+void
+test_callbacks_are_only_called_once_even_if_multiple_dependencies_change(void)
```

Comment:
I probably should take this to mean the test name is too long, but it's the name in canonical-data.json, and am unsure how to abbreviate.